File: source/math/Int2.java
package math;

import java.io.Serializable;

public class Int2 implements Serializable {
    public int x;
    public int y;

    public Int2() {
        set(0);
    }

    public Int2(int a) {
        set(a);
    }

    public Int2(int[] data) {
        set(data);
    }

    public Int2(int x, int y) {
        set(x, y);
    }

    public Int2(Int2 v) {
        set(v);
    }

    public Int2(Float2 v) {
        set(v);
    }

    // Setters
    public void set(int x, int y) {
        this.x = x; this.y = y;
    }

    public void set(int a) {
        set(a, a);
    }

    public void set(int[] data) {
        set(data[0], data[1]);
    }

    public void set(Int2 v) {
        set(v.x, v.y);
    }

    public void set(Float2 v) {
        set((int) v.x, (int) v.y);
    }

    // Getters
    public int[] array() {
        return new int[] { x, y };
    }

    // Math
    public Int2 add(Int2 v) {
        return new Int2(x + v.x, y + v.y);
    }

    public Int2 subtract(Int2 v) {
        return new Int2(x - v.x, y - v.y);
    }

    public Int2 multiply(int a) {
        return new Int2(x * a, y * a);
    }

    public Int2 multiply(Int2 v) {
        return new Int2(x * v.x, y * v.y);
    }

    public Int2 divide(int a) {
        return new Int2(x / a, y / a);
    }

    public Int2 divide(Int2 v) {
        return new Int2(x / v.x, y / v.y);
    }

    public boolean equals(Int2 v) {
        return x == v.x && y == v.y;
    }

    public Int2 negate() {
        return multiply(-1);
    }

    public Int2 absolute() {
        return new Int2(Math.abs(x), Math.abs(y));
    }

    public static Int2 getPosX() {
        return new Int2(1, 0);
    }
    public static Int2 getNegX() {
        return new Int2(-1, 0);
    }
    public static Int2 getPosY() {
        return new Int2(0, 1);
    }
    public static Int2 getNegY() {
        return new Int2(0, -1);
    }

    public String toString() {
        return "(" + x + ", " + y + ")";
    }
}

File: source/math/Int2.kt
package math

import java.io.Serializable

class Int2 : Serializable {
    var x = 0
    var y = 0

    constructor(x: Int, y: Int) {
        this.x = x
        this.y = y
    }

    constructor() : this(0, 0)

    constructor(a: Int) : this(a, a)

    constructor(data: IntArray) : this(data[0], data[1])

    constructor(v: Int2) : this(v.x, v.y)

    constructor(v: Float2) : this(v.x.toInt(), v.y.toInt())

    // Getters
    val array: IntArray
        get() = intArrayOf(x, y)

    // Math
    operator fun plus(v: Int2): Int2 {
        return Int2(x + v.x, y + v.y)
    }

    operator fun minus(v: Int2): Int2 {
        return Int2(x - v.x, y - v.y)
    }

    operator fun times(a: Int): Int2 {
        return Int2(x * a, y * a)
    }

    operator fun div(a: Int): Int2 {
        return Int2(x / a, y / a)
    }

    override fun equals(other: Any?): Boolean {
        if (other is Int2) {
            return x == other.x && y == other.y
        }
        return false
    }

    operator fun unaryMinus(): Int2 {
        return this * -1
    }

    override fun toString(): String {
        return "($x, $y)"
    }

    override fun hashCode(): Int {
        var result = x
        result = 31 * result + y
        return result
    }

    companion object {
        val posX: Int2
            get() = Int2(1, 0)
        val negX: Int2
            get() = Int2(-1, 0)
        val posY: Int2
            get() = Int2(0, 1)
        val negY: Int2
            get() = Int2(0, -1)
    }
}